File: phases/17-testing-and-verification/15-symbolic-execution-klee-angr/code/main.c
#include <stdio.h>

const char *target(int x, int y) {
    if (x > 3) {
        if (y == x + 1) return "path_A";
        if (y < 0) return "path_B";
    } else {
        if (x + y == 0) return "path_C";
    }
    return "path_D";
}

int main(void) {
    printf("%s\n", target(4, 5));
    printf("%s\n", target(4, -1));
    printf("%s\n", target(2, -2));
    printf("%s\n", target(2, 2));
    return 0;
}

File: phases/17-testing-and-verification/15-symbolic-execution-klee-angr/code/main.py
#!/usr/bin/env python3
from __future__ import annotations


def target(x: int, y: int) -> str:
    if x > 3:
        if y == x + 1:
            return "path_A"
        if y < 0:
            return "path_B"
    else:
        if x + y == 0:
            return "path_C"
    return "path_D"


def find_witness(name: str) -> tuple[int, int] | None:
    for x in range(-5, 8):
        for y in range(-5, 8):
            if target(x, y) == name:
                return (x, y)
    return None


def main() -> None:
    for p in ["path_A", "path_B", "path_C", "path_D"]:
        w = find_witness(p)
        print(p, w)


if __name__ == "__main__":
    main()
